Strange behaviour when trying to index geo_shape polygon over dateline

<!--

** Please read the guidelines below. **

Issues that do not follow these guidelines are likely to be closed.

1.  GitHub is reserved for bug reports and feature requests. The best place to
    ask a general question is at the Elastic [forums](https://discuss.elastic.co).
    GitHub is not the place for general questions.

2.  Is this bug report or feature request for a supported OS? If not, it
    is likely to be closed.  See https://www.elastic.co/support/matrix#show_os

3.  Please fill out EITHER the feature request block or the bug report block
    below, and delete the other block.

-->


<!-- Bug report -->

**Elasticsearch version** (`bin/elasticsearch --version`): 5.5.1

**Plugins installed**: []

**JVM version** (`java -version`): 1.8.0_141

**OS version** (`uname -a` if on a Unix-like system): GNU/Linux

**Description of the problem including expected versus actual behavior**:

When trying to index a geo_shape which crosses the dateline am getting intersection errors. 80% of the issues were solved by subtracting a small amount (1e10^-14) from any value which lies directly on +/-180°. There were no intersections here either but ES seemed to not like the value 180.0.  There are still a few shapes which won't go in. I have investigated to see if there are any intersections and can't find any. Translating the shape to meridian using +180 on -ve values and -180 on +ve values allows it to go in so it is solely a dateline issue. 

**Steps to reproduce**:

Please include a *minimal* but *complete* recreation of the problem, including
(e.g.) index creation, mappings, settings, query etc.  The easier you make for
us to reproduce it, the more likely that somebody will take the time to look at it.

 1. PUT /test-index
```json
{
   "mappings": {
      "geo_metadata": {
         "_all": {
            "enabled": true
         },
         "properties": {
            "spatial": {
               "properties": {
                  "geometries": {
                     "properties": {
                        "search": {
                           "type": "geo_shape",
                           "tree": "quadtree",
                           "precision": "8km"
                        }
                     }
                  }
               }
            }
         }
      }
   }
}

```
 2. PUT /test-index/geo_metadata/1
```json
{
        "spatial": {
            "geometries": {
                "search":{
                    "coordinates": [[[179.99999999999, -72.661870459646082], [-179.92242898523259, -72.614739463963019], [-179.70509673080812, -72.48077664056396], [-179.49092755066465, -72.346619319446333], [-179.27982991402556, -72.212217663440484], [-179.0801618967584, -72.082942320949627], [-179.91362724448533, -72.077340115037259], [179.99999999999, -72.595885626443078], [179.98777151047784, -72.669300308135334], [179.99999999999, -72.661870459646082]]],
                    "orientation": "counterclockwise",
                    "type": "polygon"
                }
            }
        }
}

```

The error message is generated inside JTS validation that fails on `LINEARRING (-179.92242898523259 -72.61473946396302, -179.70509673080812 -72.48077664056396, -179.49092755066465 -72.34661931944633, -179.27982991402556 -72.21221766344048, -179.0801618967584 -72.08294232094963, -179.91362724448533 -72.07734011503726, 179.99999999999 -72.59588562644308, 179.98777151047784 -72.66930030813533, 179.99999999999 -72.66187045964608, -179.92242898523259 -72.61473946396302)` that we pass to it, which is indeed self-intersecting if you treat it clockwise. Since this code path is going away as a result of #32039, it makes sense to wait for it. 

Saying that I tried this shape in #35320 and got the error:

```
[elasticsearch] Caused by: java.lang.IllegalArgumentException: Unable to Tessellate shape [[-72.61473946396302, -179.92242898523259] [-72.48077664056396, -179.70509673080812] [-72.34661931944633, -179.49092755066465] [-72.21221766344048, -179.27982991402556] [-72.08294232094963, -179.0801618967584] [-72.07734011503726, -179.91362724448533] [-72.59588562644308, 179.99999999999] [-72.66930030813533, 179.98777151047784] [-72.66187045964608, 179.99999999999] [-72.61473946396302, -179.92242898523259] ]. Possible malformed shape detected.
[elasticsearch]         at org.apache.lucene.geo.Tessellator.tessellate(Tessellator.java:115) ~[lucene-sandbox-8.0.0-snapshot-61e448666d.jar:8.0.0-snapshot-61e448666d 61e448666d56c20c37dfb825ec14ffe697d000ff - nhat - 2018-12-11 18:10:56]
[elasticsearch]         at org.apache.lucene.document.LatLonShape.createIndexableFields(LatLonShape.java:72) ~[lucene-sandbox-8.0.0-snapshot-61e448666d.jar:8.0.0-snapshot-61e448666d 61e448666d56c20c37dfb825ec14ffe697d000ff - nhat - 2018-12-11 18:10:56]
[elasticsearch]         at org.elasticsearch.index.mapper.GeoShapeFieldMapper.indexShape(GeoShapeFieldMapper.java:134) ~[elasticsearch-7.0.0-SNAPSHOT.jar:7.0.0-SNAPSHOT]
[elasticsearch]         at org.elasticsearch.index.mapper.GeoShapeFieldMapper.parse(GeoShapeFieldMapper.java:117) ~[elasticsearch-7.0.0-SNAPSHOT.jar:7.0.0-SNAPSHOT]
```
So, it looks like the issue migrated to the new system as well somehow.

--- a/server/src/main/java/org/elasticsearch/common/geo/GeoPolygonDecomposer.java
+++ b/server/src/main/java/org/elasticsearch/common/geo/GeoPolygonDecomposer.java
@@ -154,7 +154,8 @@ public class GeoPolygonDecomposer {
         }
         if (signedArea == 0) {
             // Points are collinear or self-intersection
-            throw new InvalidShapeException("Cannot determine orientation: signed area equal to 0");
+            throw new InvalidShapeException("Cannot determine orientation: signed area equal to 0." +
+                " Points are collinear or polygon self-intersects.");
         }
         boolean orientation = signedArea < 0;
 

--- a/server/src/main/java/org/elasticsearch/common/geo/GeoPolygonDecomposer.java
+++ b/server/src/main/java/org/elasticsearch/common/geo/GeoPolygonDecomposer.java
@@ -36,7 +36,6 @@ import java.util.List;
 import java.util.Set;
 import java.util.concurrent.atomic.AtomicBoolean;
 
-import static org.apache.lucene.geo.GeoUtils.orient;
 import static org.elasticsearch.common.geo.GeoUtils.normalizeLat;
 import static org.elasticsearch.common.geo.GeoUtils.normalizeLon;
 
@@ -190,45 +189,17 @@ public class GeoPolygonDecomposer {
      * @return whether the points are clockwise (true) or anticlockwise (false)
      */
     private static boolean getOrientation(Point[] points, int offset, int length) {
-        // calculate the direction of the points: find the southernmost point
-        // and check its neighbors orientation.
-
-        final int top = top(points, offset, length);
-        final int prev = (top + length - 1) % length;
-        final int next = (top + 1) % length;
-
-        final int determinantSign = orient(
-            points[offset + prev].getX(), points[offset + prev].getY(),
-            points[offset + top].getX(), points[offset + top].getY(),
-            points[offset + next].getX(), points[offset + next].getY());
-
-        if (determinantSign == 0) {
-            // Points are collinear, but `top` is not in the middle if so, so the edges either side of `top` are intersecting.
-            throw new InvalidShapeException("Cannot determine orientation: edges adjacent to ("
-                + points[offset + top].getX() + "," + points[offset + top].getY() + ") coincide");
+        // calculate the signed area of the points:
+        double determinantSign = 0;
+        for (int i = offset; i < offset + length; i++ ) {
+            determinantSign += points[i].getX() * points[i + 1].getY() - points[i].getY() * points[i + 1].getX();
         }
-
+        if (determinantSign == 0) {
+            // Points are collinear or self-intersection
+            throw new InvalidShapeException("Cannot determine orientation: signed area equal to 0");
+        }
         return determinantSign < 0;
     }
-
-    /**
-     * @return the (offset) index of the point that is furthest west amongst
-     * those points that are the furthest south in the set.
-     */
-    private static int top(Point[] points, int offset, int length) {
-        int top = 0; // we start at 1 here since top points to 0
-        for (int i = 1; i < length; i++) {
-            if (points[offset + i].getY() < points[offset + top].getY()) {
-                top = i;
-            } else if (points[offset + i].getY() == points[offset + top].getY()) {
-                if (points[offset + i].getX() < points[offset + top].getX()) {
-                    top = i;
-                }
-            }
-        }
-        return top;
-    }
-
 
     private static double[] range(Point[] points, int offset, int length) {
         double minX = points[0].getX();

--- a/server/src/main/java/org/elasticsearch/common/geo/GeoPolygonDecomposer.java
+++ b/server/src/main/java/org/elasticsearch/common/geo/GeoPolygonDecomposer.java
@@ -144,17 +144,27 @@ public class GeoPolygonDecomposer {
      */
     private static Edge[] ring(int component, boolean direction, boolean handedness,
                         Point[] points, int offset, Edge[] edges, int toffset, int length, final AtomicBoolean translated) {
-
-        boolean orientation = getOrientation(points, offset, length);
+        double signedArea = 0;
+        double minX = Double.POSITIVE_INFINITY;
+        double maxX = Double.NEGATIVE_INFINITY;
+        for (int i = offset; i < offset + length; i++) {
+            signedArea += points[i].getX() * points[i + 1].getY() - points[i].getY() * points[i + 1].getX();
+            minX = Math.min(minX, points[i].getX());
+            maxX = Math.max(maxX, points[i].getX());
+        }
+        if (signedArea == 0) {
+            // Points are collinear or self-intersection
+            throw new InvalidShapeException("Cannot determine orientation: signed area equal to 0");
+        }
+        boolean orientation = signedArea < 0;
 
         // OGC requires shell as ccw (Right-Handedness) and holes as cw (Left-Handedness)
         // since GeoJSON doesn't specify (and doesn't need to) GEO core will assume OGC standards
         // thus if orientation is computed as cw, the logic will translate points across dateline
         // and convert to a right handed system
 
-        // compute the bounding box and calculate range
-        double[] range = range(points, offset, length);
-        final double rng = range[1] - range[0];
+        // calculate range
+        final double rng = maxX - minX;
         // translate the points if the following is true
         //   1.  shell orientation is cw and range is greater than a hemisphere (180 degrees) but not spanning 2 hemispheres
         //       (translation would result in a collapsed poly)
@@ -183,46 +193,6 @@ public class GeoPolygonDecomposer {
                 points[i] = new Point(points[i].getX() + 2 * DATELINE, points[i].getY());
             }
         }
-    }
-
-    /**
-     * @return whether the points are clockwise (true) or anticlockwise (false)
-     */
-    private static boolean getOrientation(Point[] points, int offset, int length) {
-        // calculate the signed area of the points:
-        double determinantSign = 0;
-        for (int i = offset; i < offset + length; i++ ) {
-            determinantSign += points[i].getX() * points[i + 1].getY() - points[i].getY() * points[i + 1].getX();
-        }
-        if (determinantSign == 0) {
-            // Points are collinear or self-intersection
-            throw new InvalidShapeException("Cannot determine orientation: signed area equal to 0");
-        }
-        return determinantSign < 0;
-    }
-
-    private static double[] range(Point[] points, int offset, int length) {
-        double minX = points[0].getX();
-        double maxX = minX;
-        double minY = points[0].getY();
-        double maxY = minY;
-        // compute the bounding coordinates (@todo: cleanup brute force)
-        for (int i = 1; i < length; ++i) {
-            Point point = points[offset + i];
-            if (point.getX() < minX) {
-                minX = point.getX();
-            }
-            if (point.getX() > maxX) {
-                maxX = point.getX();
-            }
-            if (point.getY() < minY) {
-                minY = point.getY();
-            }
-            if (point.getY() > maxY) {
-                maxY = point.getY();
-            }
-        }
-        return new double[]{minX, maxX, minY, maxY};
     }
 
     private static int merge(Edge[] intersections, int offset, int length, Edge[] holes, int numHoles) {

--- a/server/src/main/java/org/elasticsearch/common/geo/builders/PolygonBuilder.java
+++ b/server/src/main/java/org/elasticsearch/common/geo/builders/PolygonBuilder.java
@@ -46,8 +46,6 @@ import java.util.Locale;
 import java.util.Objects;
 import java.util.concurrent.atomic.AtomicBoolean;
 
-import static org.apache.lucene.geo.GeoUtils.orient;
-
 /**
  * The {@link PolygonBuilder} implements the groundwork to create polygons. This contains
  * Methods to wrap polygons at the dateline and building shapes from the data held by the
@@ -712,43 +710,18 @@ public class PolygonBuilder extends ShapeBuilder<JtsGeometry, org.elasticsearch.
      * @return whether the points are clockwise (true) or anticlockwise (false)
      */
     private static boolean getOrientation(Coordinate[] points, int offset, int length) {
-        // calculate the direction of the points: find the southernmost point
-        // and check its neighbors orientation.
-
-        final int top = top(points, offset, length);
-        final int prev = (top + length - 1) % length;
-        final int next = (top + 1) % length;
-
-        final int determinantSign = orient(
-            points[offset + prev].x, points[offset + prev].y,
-            points[offset + top].x, points[offset + top].y,
-            points[offset + next].x, points[offset + next].y);
+        // calculate the signed area of the points:
+        double determinantSign = 0;
+        for (int i = offset; i < offset + length; i++) {
+            determinantSign += points[i].x * points[i + 1].y - points[i].y * points[i + 1].x;
+        }
 
         if (determinantSign == 0) {
-            // Points are collinear, but `top` is not in the middle if so, so the edges either side of `top` are intersecting.
-            throw new InvalidShapeException("Cannot determine orientation: edges adjacent to ("
-                + points[offset + top].x + "," + points[offset +top].y + ") coincide");
+            // Points are collinear
+            throw new InvalidShapeException("Cannot determine orientation: signed area equal to 0");
         }
 
         return determinantSign < 0;
-    }
-
-    /**
-     * @return the (offset) index of the point that is furthest west amongst
-     * those points that are the furthest south in the set.
-     */
-    private static int top(Coordinate[] points, int offset, int length) {
-        int top = 0; // we start at 1 here since top points to 0
-        for (int i = 1; i < length; i++) {
-            if (points[offset + i].y < points[offset + top].y) {
-                top = i;
-            } else if (points[offset + i].y == points[offset + top].y) {
-                if (points[offset + i].x < points[offset + top].x) {
-                    top = i;
-                }
-            }
-        }
-        return top;
     }
 
     private static double[] range(Coordinate[] points, int offset, int length) {

--- a/server/src/main/java/org/elasticsearch/common/geo/builders/PolygonBuilder.java
+++ b/server/src/main/java/org/elasticsearch/common/geo/builders/PolygonBuilder.java
@@ -676,17 +676,27 @@ public class PolygonBuilder extends ShapeBuilder<JtsGeometry, org.elasticsearch.
      */
     private static Edge[] ring(int component, boolean direction, boolean handedness,
                                  Coordinate[] points, int offset, Edge[] edges, int toffset, int length, final AtomicBoolean translated) {
-
-        boolean orientation = getOrientation(points, offset, length);
+        double signedArea = 0;
+        double minX = Double.POSITIVE_INFINITY;
+        double maxX = Double.NEGATIVE_INFINITY;
+        for (int i = offset; i < offset + length; i++) {
+            signedArea += points[i].x * points[i + 1].y - points[i].y * points[i + 1].x;
+            minX = Math.min(minX, points[i].x);
+            maxX = Math.max(maxX, points[i].x);
+        }
+        if (signedArea == 0) {
+            // Points are collinear or self-intersection
+            throw new InvalidShapeException("Cannot determine orientation: signed area equal to 0");
+        }
+        boolean orientation = signedArea < 0;
 
         // OGC requires shell as ccw (Right-Handedness) and holes as cw (Left-Handedness)
         // since GeoJSON doesn't specify (and doesn't need to) GEO core will assume OGC standards
         // thus if orientation is computed as cw, the logic will translate points across dateline
         // and convert to a right handed system
 
-        // compute the bounding box and calculate range
-        double[] range = range(points, offset, length);
-        final double rng = range[1] - range[0];
+        // calculate range
+        final double rng = maxX - minX;
         // translate the points if the following is true
         //   1.  shell orientation is cw and range is greater than a hemisphere (180 degrees) but not spanning 2 hemispheres
         //       (translation would result in a collapsed poly)
@@ -704,47 +714,6 @@ public class PolygonBuilder extends ShapeBuilder<JtsGeometry, org.elasticsearch.
             }
         }
         return concat(component, direction ^ orientation, points, offset, edges, toffset, length);
-    }
-
-    /**
-     * @return whether the points are clockwise (true) or anticlockwise (false)
-     */
-    private static boolean getOrientation(Coordinate[] points, int offset, int length) {
-        // calculate the signed area of the points:
-        double determinantSign = 0;
-        for (int i = offset; i < offset + length; i++) {
-            determinantSign += points[i].x * points[i + 1].y - points[i].y * points[i + 1].x;
-        }
-
-        if (determinantSign == 0) {
-            // Points are collinear
-            throw new InvalidShapeException("Cannot determine orientation: signed area equal to 0");
-        }
-
-        return determinantSign < 0;
-    }
-
-    private static double[] range(Coordinate[] points, int offset, int length) {
-        double minX = points[0].x;
-        double maxX = points[0].x;
-        double minY = points[0].y;
-        double maxY = points[0].y;
-        // compute the bounding coordinates (@todo: cleanup brute force)
-        for (int i = 1; i < length; ++i) {
-            if (points[offset + i].x < minX) {
-                minX = points[offset + i].x;
-            }
-            if (points[offset + i].x > maxX) {
-                maxX = points[offset + i].x;
-            }
-            if (points[offset + i].y < minY) {
-                minY = points[offset + i].y;
-            }
-            if (points[offset + i].y > maxY) {
-                maxY = points[offset + i].y;
-            }
-        }
-        return new double[] {minX, maxX, minY, maxY};
     }
 
     /**

--- a/server/src/test/java/org/elasticsearch/common/geo/GeometryIndexerTests.java
+++ b/server/src/test/java/org/elasticsearch/common/geo/GeometryIndexerTests.java
@@ -370,6 +370,19 @@ public class GeometryIndexerTests extends ESTestCase {
         assertThat(e.getMessage(), not(containsString("NaN")));
     }
 
+    public void testCrossingDateline() {
+        Polygon polygon = new Polygon(new LinearRing(
+            new double[]{170, -170, -170, 170, 170}, new double[]{-10, -10, 10, 10, -10}
+        ));
+        Geometry geometry = indexer.prepareForIndexing(polygon);
+        assertTrue(geometry instanceof MultiPolygon);
+        polygon = new Polygon(new LinearRing(
+            new double[]{180, -170, -170, 170, 180}, new double[]{-10, -5, 15, -15, -10}
+        ));
+        geometry = indexer.prepareForIndexing(polygon);
+        assertTrue(geometry instanceof MultiPolygon);
+    }
+
     private XContentBuilder polygon(Boolean orientation, double... val) throws IOException {
         XContentBuilder pointGeoJson = XContentFactory.jsonBuilder().startObject();
         {

--- a/server/src/test/java/org/elasticsearch/common/geo/ShapeBuilderTests.java
+++ b/server/src/test/java/org/elasticsearch/common/geo/ShapeBuilderTests.java
@@ -248,7 +248,7 @@ public class ShapeBuilderTests extends ESTestCase {
                     .coordinate(-40.0, -50.0)
                     .coordinate(40.0, -50.0).close());
         Exception e = expectThrows(InvalidShapeException.class, () -> newPolygon.buildS4J());
-        assertThat(e.getMessage(), containsString("Self-intersection at or near point (0.0"));
+        assertThat(e.getMessage(), containsString("Cannot determine orientation: signed area equal to 0"));
     }
 
     /** note: only supported by S4J at the moment */

--- a/server/src/test/java/org/elasticsearch/common/geo/builders/PolygonBuilderTests.java
+++ b/server/src/test/java/org/elasticsearch/common/geo/builders/PolygonBuilderTests.java
@@ -24,6 +24,7 @@ import org.elasticsearch.common.geo.builders.ShapeBuilder.Orientation;
 import org.elasticsearch.test.geo.RandomShapeGenerator;
 import org.elasticsearch.test.geo.RandomShapeGenerator.ShapeType;
 import org.locationtech.spatial4j.exception.InvalidShapeException;
+import org.locationtech.spatial4j.shape.jts.JtsGeometry;
 
 import java.io.IOException;
 
@@ -159,6 +160,18 @@ public class PolygonBuilderTests extends AbstractShapeBuilderTestCase<PolygonBui
         PolygonBuilder pb = new PolygonBuilder(new CoordinatesBuilder()
             .coordinate(0.0, 0.0).coordinate(1.0, 1.0).coordinate(-1.0, -1.0).close());
         InvalidShapeException e = expectThrows(InvalidShapeException.class, pb::buildS4J);
-        assertEquals("Cannot determine orientation: edges adjacent to (-1.0,-1.0) coincide", e.getMessage());
+        assertEquals("Cannot determine orientation: signed area equal to 0", e.getMessage());
+    }
+
+    public void testCrossingDateline() {
+        PolygonBuilder pb = new PolygonBuilder(new CoordinatesBuilder()
+            .coordinate(170, -10).coordinate(-170, -10).coordinate(-170, 10).coordinate(170, 10).coordinate(170, -10));
+        JtsGeometry geometry = pb.buildS4J();
+
+        assertTrue(geometry.getGeom() instanceof org.locationtech.jts.geom.MultiPolygon);
+        pb = new PolygonBuilder(new CoordinatesBuilder()
+            .coordinate(180, -10).coordinate(-170, -5).coordinate(-170, 15).coordinate(170, -15).coordinate(180, -10));
+        geometry = pb.buildS4J();
+        assertTrue(geometry.getGeom() instanceof org.locationtech.jts.geom.MultiPolygon);
     }
 }

